src/ips.rs

```diff
@@ -251,29 +223,84 @@ impl Ips {
                 candidates.sort_by(|a, b| {
-                    state.nodes[a.index]
+                    working_state.nodes[a.index]
                         .betweenness
-                        .partial_cmp(&state.nodes[b.index].betweenness)
+                        .partial_cmp(&working_state.nodes[b.index].betweenness)
                         .unwrap()
                 });
 
                 for peer in candidates.iter().take(peers_to_add_count as usize) {
                     peer_list_entry.list.push(peer.ip);
                 }
             }
+            working_state.peer_list.push(peer_list_entry);
+        }
 
-            // Do not compute factors one more time after every single peerlist addition. At least
-            // for now, when computing factors is very expensive (especially betweenness and closeness).
-            // Re-calculating it after each node for whole graph would take too long.
-            // TODO(asmie): recalculate some factors (like islands) after each node to check if graph is still connected
+        // TODO(asmie): recalculate and compare factors to check if network is going better
 
-            peer_list.push(peer_list_entry);
-        }
-        peer_list
+        working_state.peer_list
     }
 
     // Helper functions
 
+    /// Generate state for IPS
+    fn generate_state(&self, nodes: &[Node]) -> IpsState {
+        let mut ips_state = IpsState {
+            nodes: nodes.to_vec(),
+            ..Default::default()
+        };
+
+        let mut graph = construct_graph(nodes);
+        let betweenness = graph.betweenness_centrality();
+        let closeness = graph.closeness_centrality();
+
+        // Recalculate factors with new graph
+        for node in ips_state.nodes.iter_mut() {
+            let ip = IpAddr::from_str(&node.ip).unwrap();
+            node.betweenness = *betweenness.get(&ip).expect("can't fetch betweenness");
+            node.closeness = *closeness.get(&ip).expect("can't fetch closeness");
+        }
+
+        ips_state.degrees = graph.degree_centrality();
+        ips_state.eigenvalues = graph.eigenvalue_centrality();
+
+        ips_state.degree_factors = NormalizationFactors::determine(
+            &ips_state.degrees.values().cloned().collect::<Vec<u32>>(),
+        );
+
+        ips_state.eigenvector_factors = NormalizationFactors::determine(
+            &ips_state
+                .eigenvalues
+                .values()
+                .cloned()
+                .collect::<Vec<f64>>(),
+        );
+
+        let betweenness = &nodes.iter().map(|n| n.betweenness).collect::<Vec<f64>>();
+        ips_state.betweenness_factors = NormalizationFactors::determine(betweenness);
+
+        let closeness = &nodes.iter().map(|n| n.closeness).collect::<Vec<f64>>();
+        ips_state.closeness_factors = NormalizationFactors::determine(closeness);
+
+        ips_state.peer_list = Peer::generate_peerlist(nodes);
```

Comment:
I was not sure if this is generating only one peer list or not (from the function signature)
Maybe:
```suggestion
        ips_state.peer_list = Peer::generate_peerlists(nodes);
```
or maybe `generate_all_peerlists`

Reply:
Done.